pkg/alertmanager/multitenant_test.go

```diff
@@ -368,3 +395,456 @@ receivers:
+				desc.AddIngester("alertmanager-1", "127.0.0.1", "", ring.Tokens{1, 2, 3}, ring.ACTIVE, registeredAt)
+				desc.AddIngester("alertmanager-2", "127.0.0.2", "", ring.Tokens{4, 5, 6}, ring.ACTIVE, registeredAt)
+			},
+			updateRing: func(desc *ring.Desc) {
+				instance := desc.Ingesters["alertmanager-1"]
+				instance.Timestamp = time.Now().Add(time.Second).Unix()
+				desc.Ingesters["alertmanager-1"] = instance
+			},
+			expected: false,
+		},
+		{
+			name: "should NOT sync when an instance is auto-forgotten in the ring but was already unhealthy in the previous state",
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester("alertmanager-1", "127.0.0.1", "", ring.Tokens{1, 2, 3}, ring.ACTIVE, registeredAt)
+				desc.AddIngester("alertmanager-2", "127.0.0.2", "", ring.Tokens{4, 5, 6}, ring.ACTIVE, registeredAt)
+
+				instance := desc.Ingesters["alertmanager-2"]
+				instance.Timestamp = time.Now().Add(-time.Hour).Unix()
+				desc.Ingesters["alertmanager-2"] = instance
+			},
+			updateRing: func(desc *ring.Desc) {
+				desc.RemoveIngester("alertmanager-2")
+			},
+			expected: false,
+		},
+	}
+
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			amConfig := mockAlertmanagerConfig(t)
+			amConfig.ShardingEnabled = true
+			amConfig.ShardingRing.RingCheckPeriod = 100 * time.Millisecond
+			amConfig.PollInterval = time.Hour // Don't trigger the periodic check.
+
+			ringStore := consul.NewInMemoryClient(ring.GetCodec())
+			mockStore := &mockAlertStore{
+				configs: map[string]alerts.AlertConfigDesc{},
+			}
+
+			reg := prometheus.NewPedanticRegistry()
+			am, err := createMultitenantAlertmanager(amConfig, nil, nil, mockStore, ringStore, log.NewNopLogger(), reg)
+			require.NoError(t, err)
+
+			require.NoError(t, ringStore.CAS(ctx, RingKey, func(in interface{}) (interface{}, bool, error) {
+				ringDesc := ring.GetOrCreateRingDesc(in)
+				tt.setupRing(ringDesc)
+				return ringDesc, true, nil
+			}))
+
+			require.NoError(t, services.StartAndAwaitRunning(ctx, am))
+			defer services.StopAndAwaitTerminated(ctx, am) //nolint:errcheck
+
+			// Make sure the initial sync happened.
+			regs := util.NewUserRegistries()
+			regs.AddUserRegistry("test", reg)
+			metrics := regs.BuildMetricFamiliesPerUser()
+			assert.Equal(t, float64(1), metrics.GetSumOfCounters("cortex_alertmanager_sync_configs_total"))
+
+			// Change the ring topology.
+			require.NoError(t, ringStore.CAS(ctx, RingKey, func(in interface{}) (interface{}, bool, error) {
+				ringDesc := ring.GetOrCreateRingDesc(in)
+				tt.updateRing(ringDesc)
+				return ringDesc, true, nil
+			}))
+
+			// Assert if we expected a sync or not.
+			if tt.expected {
+				test.Poll(t, time.Second, float64(2), func() interface{} {
+					metrics := regs.BuildMetricFamiliesPerUser()
+					return metrics.GetSumOfCounters("cortex_alertmanager_sync_configs_total")
+				})
+			} else {
+				time.Sleep(250 * time.Millisecond)
```

Comment:
Why this sleep? We try to avoid sleeps in tests because are a source of flakyness. Can be converted into a `test.Poll()`?

Reply:
I tried originally, but couldn't. The problem here is that we have no way of saying "wait for at least X polls before checking". 250ms is 2.5x the time needed for sync to trigger so hopefully this doesn't become a flake for a while..

The assertion below is comparing that the value did not change because only the initial sync happened. Hence a value of 1.